pypi_resource/out.py

```diff
@@ -81,7 +80,10 @@ def out(srcdir, input):
         raise NamesValidationError(
             f"Different names for package ({package_name}) and input ({input_name}). If this is intentional, you can configure `name_must_match` to `false`."
         )
-    if not is_canonical(version):
+
+    try:
+        pipio.Version(str(version))
+    except pipio.InvalidVersion:
         raise VersionValidationError(
             f"Version {version} string is not compliant with PEP 440 versioning convention"
```

Comment:
I think that "string" is redundant and can be omitted.

But the most important question is, since we removed the pep440 module, whether the explanation "is not compliant with PEP 440" is still correct.

I wonder also what are the contents of pipio.InvalidVersion, does it contain useful hints for the user?

Reply:
And this makes me realize that the code is maybe "wrong". 

Current behavior of the resource (before this PR): Its forbidden to upload to any repository (public or private) any version with local identifier. Only public identifiers according to PEP440 are allowed.
Behavior after this PR: its possible to upload to any repository any PEP440 acceptable version (even with local identifier). 

But if I understand correctly uploading to PyPi a version with local identifier is wrong and should be rejected by PyPI. I will test this now.

Reply:
Consider also:

> I wonder also what are the contents of pipio.InvalidVersion, does it contain useful hints for the user?

Reply:
So yes, I need to refactor this code a little bit better. 
As I suspected above, it will not be possible to push a version with local identifier to public repos like (pypi or test-pypi).
```
HTTPError: 400 Bad Request from https://test.pypi.org/legacy/          
         '1.2.3.dev10+g123abc45' is an invalid value for Version. Error: Can't  
         use PEP 440 local versions. See                                        
         https://packaging.python.org/specifications/core-metadata for more     
         information. 
```

Reply:
> I wonder also what are the contents of pipio.InvalidVersion, does it contain useful hints for the user?
Simple test:
```
>>> Version("1.2.3.dev10-g123abc45")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/alicul/WORK/concourse-pypi-resource/.venv/lib/python3.9/site-packages/pip/_vendor/packaging/version.py", line 277, in __init__
    raise InvalidVersion("Invalid version: '{0}'".format(version))
pip._vendor.packaging.version.InvalidVersion: Invalid version: '1.2.3.dev10-g123abc45'
```
but, as you pointed in one of your comments, maybe we should modify this message to:
VersionValidationError(
            f"Version {version} string is not compliant with PEP 440 versioning convention. See https://peps.python.org/pep-0440 for more details."

Reply:
> So yes, I need to refactor this code a little bit better. As I suspected above, it will not be possible to push a version with local identifier to public repos like (pypi or test-pypi).
> 
> ```
> HTTPError: 400 Bad Request from https://test.pypi.org/legacy/          
>          '1.2.3.dev10+g123abc45' is an invalid value for Version. Error: Can't  
>          use PEP 440 local versions. See                                        
>          https://packaging.python.org/specifications/core-metadata for more     
>          information. 
> ```

Just to be sure you're not doing useless work, you should not try to block pushing local version but instead try to surface the error reported by the server properly. The reason for this is that you only know that the official Pypi block it. Any other organisation might put in place a similar rule on its own server and so you should not adjust the resource behaviour based on the target server.

Reply:
> Just to be sure you're not doing useless work, you should not try to block pushing local version but instead try to surface the error reported by the server properly. 

But what can we do about it? At this [line](https://github.com/Pix4D/concourse-pypi-resource/blob/9d6525f8eaeafaef96d348ac181c99d256610270/pypi_resource/out.py#L60) twine will try to upload the version with local identifier and it will fail (if the server forbids it). The twine will print the error message returned by the server and we see that in case of PyPi (and test PyPI) the error is super clear. I can not guess what the error msg from other servers will be. Additionally upload can fail due to many other causes, so having a try/expect around twine subprocess  to inspect the error does not make a lot of sense. Do you agree?

Reply:
If `twine`'s error message goes through and is usually clear, then that's fine for me. We should just let it go through and make sure we fail the `put` step as a consequence.

For private server, of course we cannot be sure how they will behave but if they are configured as well as Pypi then twine will provide a good error message so it's fine. If they are not configured as well then up to their maintainer to fix that.

Reply:
I agree with @odormond that this is the good approach (letting the server deal with conformance).

The error printed by Concourse should NOT have the traceback though, it gives the impression that it is an internal error / bug.

More exactly: if possible, the code should detect if the error is an HTTP code != 200 and print the payload without traceback.

For any other cases, printing the traceback is OK, it can show real bugs in the client implementations.

Reply:
You can still discard the traceback by doing:
```python
def upload_package(pkgpath, input):
  ...
  try:
    subprocess.run(twine, check=True)
  except subprocess.CalledProcessError as e:
    raise SystemExit(e.returncode)
```

That would reduce the output to:
```
Finding package to upload
Glob build/* matched files: ['/tmp/build/put/build/private_concourse_example-1.2.3.dev10+g123abc45-py3-none-any.whl']
Check that the package name = input name
Uploading /tmp/build/put/build/private_concourse_example-1.2.3.dev10+g123abc45-py3-none-any.whl version 1.2.3.dev10+g123abc45
Uploading distributions to https://test.pypi.org/legacy/
Uploading private_concourse_example-1.2.3.dev10+g123abc45-py3-none-any.whl
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4.6 kB • 00:00 • ?
25hWARNING  Error during upload. Retry with the --verbose option for more details.
ERROR    HTTPError: 400 Bad Request from https://test.pypi.org/legacy/
         '1.2.3.dev10+g123abc45' is an invalid value for Version. Error: Can't
         use PEP 440 local versions. See
         https://packaging.python.org/specifications/core-metadata for more
         information.
```

Reply:
@odormond I did not know this. Thanks. I though that I had to add a `print(e)` just before `raise SystemExit(e.returncode)`.

But now I am wondering what happens if something else go wrong i.e anything else than `subprocess.CalledProcessError`. The code would continue and it put step would be marked as success?

Reply:
Any other exception would not be caught and would propagate upward. If nothing catches them then you'll get a traceback printed and the interpreter will exit with error code 1.